[BOLT][PAC] Warn about synchronous unwind tables

bgergely0 opened the pull request:
BOLT currently ignores functions with synchronous PAuth DWARF info.
If more than 10% of functions get ignored for inconsistencies, we
should emit a warning to only use asynchronous unwind tables.

See related issue: #165215

bgergely0 committed:
[BOLT][PAC] Warn about synchronous unwind tables

BOLT currently ignores functions with synchronous PAuth DWARF info.
When more than 10% of functions get ignored for inconsistencies, we
should emit a warning to only use asynchronous unwind tables.

See also: #165215

bgergely0 committed:
[BOLT] Use opts::Verbosity in PointerAuthCFIAnalyzer

bgergely0 commented on bolt/lib/Passes/PointerAuthCFIAnalyzer.cpp:
```diff
@@ -133,11 +140,17 @@ Error PointerAuthCFIAnalyzer::runOnFunctions(BinaryContext &BC) {
 
+  if (IgnoredPercent >= 10.0)
+    BC.outs() << "BOLT-WARNING: PointerAuthCFIAnalyzer only supports "
+                 "asynchronous unwind tables.\n";
```
updated

paschalis-mpeis reviewed (commented):
Can you elaborate how the >=10% threshold for the warning was chosen (instead of emiting it in all cases)? Maybe some of these details might also fit in the design doc too , where you talk about inconsistencies.

bgergely0 commented:
As the related issue states:

> Example: when applying BOLT to llama.cpp, this is seen in the logs: Ignored 1633 functions (99.03%) because of CFI inconsistencies. The huge % of ignored functions makes BOLT useless in such cases.

I think the double-digit territory is a good boundary to start emitting this warning. After all, we don't check/don't know if this is the reason for the errors, but the higher the percentage the more likely that this is the issue. Low-percentage issues can be from older compiler versions emitting CFIs incorrectly (as I've seen while testing the work).

So the 10% is a "magic constant" and is up for discussion, I think it is a reasonable value to start emitting warnings.

paschalis-mpeis commented:
Sounds good. So BOLT doesnt know whether inconsistencies are due to sync/async unwind tables, but instead makes an assumption to help the user? Could you summarise this with a comment in code around the point you emit the warning?

Since you are on this, should it handle cases like the below? (we've started seeing these recently)
> BOLT-INFO: MarkRAStates ran on 0 functions. Ignored 0 functions (nan%) because of CFI inconsistencies

bgergely0 committed:
[BOLT] Add comment about the chosen threshold

bgergely0 committed:
[BOLT] PointerAuthCFIAnalyzer: return early if there is no work

- makes sure we do not divide by zero, to calculate the % of ignored
  functions.

peterwaller-arm reviewed (approved):
LGTM. A few minor suggestions.

bgergely0 committed:
Review nits
